cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/RunRecordCorrectorService.java

```diff
@@ -17,10 +17,170 @@
+    // Get run records in STARTING, RUNNING and SUSPENDED states that are actually not running
+    // Do it in micro batches of transactions to avoid tx timeout
+
+    Set<ProgramRunId> fixedPrograms = new HashSet<>();
+    for (ProgramRunStatus status : NOT_STOPPED_STATUSES) {
+      long startTime = 0L;
+      while (true) {
+        Map<ProgramRunId, RunRecordMeta> runs = store.getRuns(status, startTime, Long.MAX_VALUE,
+                                                              txBatchSize, createFilter(fixedPrograms));
+        LOG.trace("{} run records in {} state but are not actually running", runs.size());
+        if (runs.isEmpty()) {
+          break;
+        }
+
+        for (RunRecordMeta record : runs.values()) {
+          startTime = Math.max(startTime, RunIds.getTime(record.getPid(), TimeUnit.SECONDS));
+
+          ProgramRunId programRunId = record.getProgramRunId();
+          if (!fixedPrograms.contains(programRunId)) {
+            programStateWriter.error(
+              programRunId,
+              new Throwable("Marking run record for " + programRunId + " as failed since no running program found."));
+
+            LOG.warn("Fixed RunRecord for program run {} in {} state because it is actually not running",
+                     programRunId, record.getStatus());
+            fixedPrograms.add(programRunId);
+          }
+        }
+      }
+    }
+
+    if (fixedPrograms.isEmpty()) {
+      LOG.trace("No RunRecord found with status in {} and the program not actually running", NOT_STOPPED_STATUSES);
+    } else {
+      LOG.warn("Fixed {} RunRecords with status in {} and the program not actually running",
+               fixedPrograms.size(), NOT_STOPPED_STATUSES);
+    }
+    return fixedPrograms;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting RunRecordCorrectorService");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Stopping RunRecordCorrectorService");
+  }
+
+  /**
+   * Creates a {@link Predicate} for {@link RunRecordMeta} to be used for scanning the {@link Store} for run records
+   * that doesn't have the program actually running as determined by the {@link ProgramRuntimeService}.
+   *
+   * @param excludedIds a set of {@link ProgramRunId} that are always rejected by the filter.
+   */
+  private Predicate<RunRecordMeta> createFilter(final Set<ProgramRunId> excludedIds) {
+    final long now = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+    return new Predicate<RunRecordMeta>() {
+      @Override
+      public boolean apply(RunRecordMeta record) {
+        ProgramRunId programRunId = record.getProgramRunId();
+
+        if (excludedIds.contains(programRunId)) {
+          return false;
+        }
+
+        // runs are marked as starting before they get into the runtimeService list.
+        // Therefore, add some buffer so that we don't incorrectly 'fix' runs that are actually starting
```

Comment:
comment is confusing here, this is actually done at the end of the method. Maybe move it down there?

Reply:
Haven't look at the message... it was copied.. I think it means don't fix record right away as the program running state may not be in the `ProgramRuntimeService` yet. I'll rephase the message better.